From bbfed408710b69d56b02d4679ba62597390f0a63 Mon Sep 17 00:00:00 2001
From: Anthony Liguori <aliguori@redhat.com>
Date: Fri, 12 Aug 2011 15:38:10 +0200
Subject: [PATCH 02/15] block: enable in_use flag

RH-Author: Anthony Liguori <aliguori@redhat.com>
Message-id: <1313163503-2523-3-git-send-email-aliguori@redhat.com>
Patchwork-id: 31326
O-Subject: [RHEL6.2 qemu PATCH 02/15] block: enable in_use flag
Bugzilla: 633370
RH-Acked-by: Marcelo Tosatti <mtosatti@redhat.com>
RH-Acked-by: Kevin Wolf <kwolf@redhat.com>
RH-Acked-by: Orit Wasserman <owasserm@redhat.com>
RH-Acked-by: Christoph Hellwig <chellwig@redhat.com>

From: Marcelo Tosatti <mtosatti@redhat.com>

Set block device in use during block migration, disallow drive_del and
bdrv_truncate for in use devices.

Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>
Signed-off-by: Kevin Wolf <kwolf@redhat.com>
Signed-off-by: Anthony Liguori <aliguori@redhat.com>

Bugzilla: 633370

(cherry picked from commit 8591675f44929a9e4b5d3a5fd702a4b6d41c7903)
---
 block-migration.c |    2 ++
 block.c           |    2 ++
 blockdev.c        |    4 ++++
 3 files changed, 8 insertions(+), 0 deletions(-)

Signed-off-by: Michal Novotny <mignov@gmail.com>
---
 block-migration.c |    2 ++
 block.c           |    2 ++
 blockdev.c        |    4 ++++
 3 files changed, 8 insertions(+), 0 deletions(-)

diff --git a/block-migration.c b/block-migration.c
index 1a8dab4..fa15a99 100644
--- a/block-migration.c
+++ b/block-migration.c
@@ -236,6 +236,7 @@ static void init_blk_migration_it(void *opaque, BlockDriverState *bs)
         bmds->total_sectors = sectors;
         bmds->completed_sectors = 0;
         bmds->shared_base = block_mig_state.shared_base;
+        bdrv_set_in_use(bs, 1);
 
         block_mig_state.total_sector_sum += sectors;
 
@@ -387,6 +388,7 @@ static void blk_mig_cleanup(Monitor *mon)
 
     while ((bmds = QSIMPLEQ_FIRST(&block_mig_state.bmds_list)) != NULL) {
         QSIMPLEQ_REMOVE_HEAD(&block_mig_state.bmds_list, entry);
+        bdrv_set_in_use(bmds->bs, 0);
         qemu_free(bmds);
     }
 
diff --git a/block.c b/block.c
index 0d57e13..22da5c7 100644
--- a/block.c
+++ b/block.c
@@ -1093,6 +1093,8 @@ int bdrv_truncate(BlockDriverState *bs, int64_t offset)
         return -ENOTSUP;
     if (bs->read_only)
         return -EACCES;
+    if (bdrv_in_use(bs))
+        return -EBUSY;
     ret = drv->bdrv_truncate(bs, offset);
     if (ret == 0) {
         ret = refresh_total_sectors(bs, offset >> BDRV_SECTOR_BITS);
diff --git a/blockdev.c b/blockdev.c
index a2df451..1f7ef2f 100644
--- a/blockdev.c
+++ b/blockdev.c
@@ -792,6 +792,10 @@ int do_drive_del(Monitor *mon, const QDict *qdict, QObject **ret_data)
         qerror_report(QERR_DEVICE_NOT_FOUND, id);
         return -1;
     }
+    if (bdrv_in_use(bs)) {
+        qerror_report(QERR_DEVICE_IN_USE, id);
+        return -1;
+    }
 
     /* quiesce block driver; prevent further io */
     qemu_aio_flush();
-- 
1.7.4.4

